jetty-core/jetty-websocket/jetty-websocket-jetty-tests/src/test/java/org/eclipse/jetty/websocket/tests/ExplicitDemandTest.java

```diff
@@ -213,4 +254,23 @@ public void testServerPing() throws Exception
         frame = pingSocket.frames.get(2);
         assertThat(frame.getType(), is(Frame.Type.CLOSE));
     }
+
+    @Test
+    public void testDemandInOnOpen() throws Exception
+    {
+        URI uri = new URI("ws://localhost:" + connector.getLocalPort() + "/onOpen");
+        EventSocket clientSocket = new EventSocket();
+
+        Future<Session> connect = client.connect(clientSocket, uri);
+        Session session = connect.get(5, TimeUnit.SECONDS);
+        session.sendText("test-text", Callback.NOOP);
+
+        String received = onOpenSocket.textMessages.poll(5, TimeUnit.SECONDS);
+        assertThat(received, equalTo("test-text"));
+        onOpenSocket.onOpen.countDown();
```

Comment:
@lachlan-roberts I think this is wrong.

We do not want 2 threads to enter WebSocket endpoint event methods concurrently, and I don't think we want the same thread to re-enter different event methods either.

The behavior that I'm expecting here is that the thread first exits `onWebSocketOpen()`, and then enters `onWebSocketFrame()`.

Reply:
What about when processing a TEXT based message handler, and a PING request comes in?

Reply:
This has now been fixed in `WebSocketConnection`.

> What about when processing a TEXT based message handler, and a PING control frame comes in and needs to be send to an endpoint ping handler?

There is already logic to ensure only one thread doing the reading, parsing and delivering of the frames, so if we are delivering a TEXT message there can be no ping/pong being read/delivered simultaneously.